File: Kodlar/Soru 3/GAPasswordCracker/src/MatingPool.java
/*
 * EGE DOĞAN DURSUN - 05170000006
 * CEM ÇORBACIOĞLU - 05130000242
 * EGE ÜNİVERSİTESİ
 * MÜHENDİSLİK FAKÜLTESİ
 * BİLGİSAYAR MÜHENDİSLİĞİ BÖLÜMÜ
 * 2019-2020 BAHAR DÖNEMİ
 * İŞLEMSEL ZEKA VE DERİN ÖĞRENME DERSİ - CIDL2020-P1
 * SORU 3 - GENETİK ALGORİTMAYLA ŞİFRE ÇÖZME
 * TARİH : 12 NİSAN 2020
 * 
 */

import java.util.ArrayList;
import java.util.Random;

public class MatingPool {

	//This method selects parents from the population depending on their fitness level.
	//Basically only parents who have either the average or more than the average fitness level of the generation
	//are selected. There is also a limit on the parents that will be selected. 
	public static ArrayList<Individual> selectParentsByFitness(ArrayList<Individual> population, 
																double averageFitness, 
																int maximumParents) {
		
		ArrayList<Individual> parents = new ArrayList();
		for (int i = 0; i<population.size(); i++) {
			if(population.get(i).getFitness() >= averageFitness){
				if(parents.size() < maximumParents) {
					parents.add(population.get(i));
				}
			}
		}
		
		return parents;
	}
	
	
	//This method creates a children by randomly selected individuals from "available parent pool". Then
	//it applies a crossing over or not depending on the crossover chance.
	public static ArrayList<Individual> produceOffspringsByCrossover(ArrayList<Individual> parents, 
																	int populationSize, 
																	int crossoverChance){
		
		ArrayList<Individual> children = new ArrayList();
		Random r = new Random();
		int chromoLength = parents.get(0).getChromosomeLength();
		
		for(int i = 0; i<populationSize; i++) {
			
			int coDie = r.nextInt(100)+1;
			String parent1 = parents.get(r.nextInt(parents.size())).getChromosome();
			String parent2 = parents.get(r.nextInt(parents.size())).getChromosome();
			String childChromosome;
			
			if (coDie < crossoverChance){
				
				int coIndex = r.nextInt(chromoLength);
				String chromoPiece1 = parent1.substring(0, coIndex);
				String chromoPiece2 = parent2.substring(coIndex);
				
				childChromosome = chromoPiece1 + chromoPiece2;
			
			}
			else {
				
				int fomDie = r.nextInt(2);
				if(fomDie == 0) {
					
					childChromosome = parent1;
				}
				else {
					childChromosome = parent2;
				}
			}
			
			Individual child = new Individual(chromoLength);
			child.setChromosome(childChromosome);
			children.add(child);
			
		}
		
		return children;
		
	}
}

File: Kodlar/Soru 3/GATravellingSalesmanProblem/MatingPool.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
EGE DOĞAN DURSUN - 05170000006
CEM ÇORBACIOĞLU - 05130000242
EGE ÜNİVERSİTESİ
MÜHENDİSLİK FAKÜLTESİ
BİLGİSAYAR MÜHENDİSLİĞİ BÖLÜMÜ
2019-2020 BAHAR DÖNEMİ
İŞLEMSEL ZEKA VE DERİN ÖĞRENME DERSİ  - CIDL2020-P1
PROJE 1 - SORU 3 - ALTERNATİF 3: GENETİK ALGORİTMALARLA TRAVELLING SALESMAN PROBLEMİ ÇÖZÜMÜ
TARİH : 13 NİSAN 2020
"""

import pandas as pd
import random
import numpy as np


#Select individuals from the ranked(sorted) population for creating the mating pool 
def selection(pop_ranked, elite_pop_size):
    selection_results = []
    df = pd.DataFrame(np.array(pop_ranked), columns=["Index","Fitness"])
    df['cum_sum'] = df.Fitness.cumsum()
    df['cum_perc'] = 100*df.cum_sum/df.Fitness.sum()
    
    for i in range(0, elite_pop_size):
        selection_results.append(pop_ranked[i][0])
    for i in range(0, len(pop_ranked) - elite_pop_size):
        pick = 100*random.random()
        for i in range(0, len(pop_ranked)):
            if pick <= df.iat[i,3]:
                selection_results.append(pop_ranked[i][0])
                break
    return selection_results
                
 
#Method for preparing the mating pool               
def mating_pool(population, selection_results):
    matingpool = []
    for i in range(0, len(selection_results)):
        index = selection_results[i]
        matingpool.append(population[index])
    return matingpool


#Breeding the parents
def breed(parent1, parent2):
    child = []
    child_P1 = []
    child_P2 = []
    
    gene_A = int(random.random() * len(parent1))
    gene_B = int(random.random() * len(parent1))
    
    start_gene = min(gene_A, gene_B)
    end_gene = max(gene_A, gene_B)

    for i in range(start_gene, end_gene):
        child_P1.append(parent1[i])
        
    child_P2 = [item for item in parent2 if item not in child_P1]

    child = child_P1 + child_P2
    return child


#Breeding method for the population
def breed_population(matingpool, elite_pop_size):
    children = []
    length = len(matingpool) - elite_pop_size
    pool = random.sample(matingpool, len(matingpool))

    for i in range(0,elite_pop_size):
        children.append(matingpool[i])
    
    for i in range(0, length):
        child = breed(pool[i], pool[len(matingpool)-i-1])
        children.append(child)
    return children
    
    